### PR TITLE
fix(api): workflow as code improvements

### LINKED ISSUE #3239: ascode: yml with hook
when we import a workflow as code, CDS creates automatically 1 repository webhook. No need to explicitly add the repowebhook in yml.

But, if user add a repoWebhook as yml below, we should have 2 repowebhook: one for the ascode, another for the user repowebhook. Actually, there is only once visible on UI (same uid) and two webhooks with same uuid on repo manager.

```
name: thewf
version: v1.0
workflow:
  ci:
    pipeline: ci
    application: theapp
    payload:
      git.author: ""
      git.branch: master
      git.hash: ""
      git.hash.before: ""
      git.message: ""
      git.repository: FOO/bar
hooks:
  ci:
    - type: RepositoryWebHook
metadata:
  default_tags: git.branch,git.author,git.hash
```

### PR DESCRIPTION
**Pull request by fsamin**
1. Description
* Since a hook is mandatory for workflow as code, do not export it
* Do not export default VCS payload
1. Related issues
* closes #3239
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report it#6418.0 ✘  
* w-it
  * w-it ✘  
Unit Tests Report
* Platform model admin [tests/cli_admin_platform_model.yml]




* Application Command TestSuite [tests/cli_application.yml]









* Action Command TestSuite with cdsctl [tests/clictl_action.yml]







* Broadcast Command TestSuite [tests/clictl_admin_broadcast.yml]


* Plugins Command TestSuite with cdsctl [tests/clictl_admin_plugin_and_platforms.yml]










* Action Command TestSuite with cdsctl [tests/clictl_admin_services.yml]


* Application Command TestSuite with CDS ctl [tests/clictl_application.yml]









* Environement Command TestSuite with CDS ctl [tests/clictl_environment.yml]









* Group Command TestSuite [tests/clictl_group.yml]












* Pipeline Command TestSuite with CDS ctl [tests/clictl_pipeline.yml]









* Project Command TestSuite with CDS ctl [tests/clictl_project.yml]












* Token Command TestSuite with CDS ctl [tests/clictl_token.yml]







* User Command TestSuite [tests/clictl_user.yml]


* Version  Command TestSuite [tests/clictl_version.yml]


* Worker Command TestSuite [tests/clictl_worker_model.yml]













* Create a simple workflow (ITSCWRKFLW3) and run it [tests/sc_workflow_create_run_join.yml]

  * Create workflow ITSCWRKFLW3 and run it ✘

* Create a simple workflow (ITSCWRKFLW1) and pull it [tests/sc_workflow_create_simple.yml]






* Create a simple workflow (ITSCWRKFLW6) and run it to test gitClone action [tests/sc_workflow_git_clone_tag.yml]



  * import workflow ✘
  * run workflow ✘
  * check workflow ✘

* Workflow (ITSCWRKFLW5) should use the worker key install command [tests/sc_workflow_key_install.yml]


  * import workflow ✘
  * run workflow ✘
  * the workflow should be successfull ✘

* Workflow (ITSCWRKFLW4) should use the worker cache command [tests/sc_workflow_run_cache.yml]


  * import workflow ✘
  * run workflow ✘
  * the workflow should be successfull ✘

* Create a simple workflow (ITSCWRKFLW2) and run it [tests/sc_workflow_run_simple.yml]







* Create a simple workflow (ITSCWRKFLW2) run it and then stop it [tests/sc_workflow_stop_simple.yml]








* Create a simple workflow (ITSCWRKFLW2), update it and run it [tests/sc_workflow_update_run.yml]

**Comment by ovh-cds** [comment]
CDS Report ut-engine#6418.0 ✘  
* Stage 1
  * Engine Test with DB ✘  
Unit Tests Report
* github.com/ovh/cds/cli


* github.com/ovh/cds/engine/api/application








* github.com/ovh/cds/engine/api/environment




* github.com/ovh/cds/engine/api/keys



* github.com/ovh/cds/engine/api/migrate


* github.com/ovh/cds/engine/api/pipeline






* github.com/ovh/cds/engine/api/plugin


* github.com/ovh/cds/engine/api/project









* github.com/ovh/cds/engine/api/secret






* github.com/ovh/cds/engine/api/user








* github.com/ovh/cds/engine/api/warning















* github.com/ovh/cds/engine/api/worker











* github.com/ovh/cds/engine/api/workflow

































  * TestImport/simple_workflow_insertion ✘
  * TestImport/simple_workflow_insertion ✘

* github.com/ovh/cds/engine/hatchery/swarm



















* github.com/ovh/cds/engine/hooks






* github.com/ovh/cds/engine/repositories


* github.com/ovh/cds/engine/vcs











* github.com/ovh/cds/engine/vcs/bitbucket


















* github.com/ovh/cds/engine/vcs/github













* github.com/ovh/cds/engine/vcs/gitlab









* github.com/ovh/cds/engine/worker






* github.com/ovh/cds/sdk


















* github.com/ovh/cds/sdk/cdsclient



* github.com/ovh/cds/sdk/exportentities




























* github.com/ovh/cds/sdk/interpolate

























* github.com/ovh/cds/sdk/luascript





* github.com/ovh/cds/sdk/namesgenerator



* github.com/ovh/cds/sdk/vcs/git

File: sdk/cdsclient/client_export.go [hunk 1]
```diff
@@ -76,12 +79,13 @@ func (c *client) WorkflowExport(projectKey, name string, exportWithPermissions b
 	return body, nil
 }
 
-func (c *client) WorkflowPull(projectKey, name string, exportWithPermissions bool) (*tar.Reader, error) {
+func (c *client) WorkflowPull(projectKey, name string, mods ...RequestModifier) (*tar.Reader, error) {
 	path := fmt.Sprintf("/project/%s/pull/workflows/%s", projectKey, name)
-	if exportWithPermissions {
+	/*if exportWithPermissions {
```
Comment (richardlt) [inline sdk/cdsclient/client_export.go]: comment

File: sdk/cdsclient/client_export.go [hunk 2]
```diff
@@ -58,12 +58,15 @@ func (c *client) EnvironmentExport(projectKey, name string, exportWithPermission
...
 }
 
-func (c *client) WorkflowExport(projectKey, name string, exportWithPermissions bool, exportFormat string) ([]byte, error) {
-	path := fmt.Sprintf("/project/%s/export/workflows/%s?format=%s", projectKey, name, exportFormat)
-	if exportWithPermissions {
-		path += "&withPermissions=true"
-	}
-	bodyReader, _, _, err := c.Stream(context.Background(), "GET", path, nil, true)
+func (c *client) WorkflowExport(projectKey, name string, mods ...RequestModifier) ([]byte, error) {
+	/*
```
Comment (richardlt) [inline sdk/cdsclient/client_export.go]: comment

File: sdk/workflow.go [hunk 3]
```diff
@@ -947,6 +948,49 @@ type WorkflowNodeContextDefaultPayloadVCS struct {
...
+	if reflect.DeepEqual(kI, kD) {
+		return true
+	}
+
+	hasKey := func(s string) bool {
+		_, has := mI[s]
+		return has
+	}
+
+	if hasKey("git.branch") &&
```
Comment (richardlt) [inline sdk/workflow.go]: return

File: sdk/workflow.go [hunk 4]
```diff
@@ -947,6 +948,49 @@ type WorkflowNodeContextDefaultPayloadVCS struct {
...
+		return true
+	}
+
+	dumper := dump.NewDefaultEncoder(nil)
+	dumper.ExtraFields.DetailedMap = false
+	dumper.ExtraFields.DetailedStruct = false
+	dumper.ExtraFields.Len = false
+	dumper.ExtraFields.Type = false
+
+	mI, _ := dumper.ToMap(i)
```
Comment (richardlt) [inline sdk/workflow.go]: Comment: // compare interface keys with default payload keys

File: sdk/workflow.go [hunk 5]
```diff
@@ -947,6 +948,49 @@ type WorkflowNodeContextDefaultPayloadVCS struct {
...
+	mD, _ := dumper.ToMap(WorkflowNodeContextDefaultPayloadVCS{})
+	var kI, kD []string
+	for k := range mI {
```
Comment (richardlt) [inline sdk/workflow.go]: if len(mI) != len(mD) { 
  return false
}
for k := range mD {
  if ! hasKey(k) {
    return false
  }
}

better than reflect.DeepEqual ?

**Pull request merged**